### PR TITLE
praxisdigital/moodle-opgaver#616 - rewrite of block_sharing_cart backup

### PR DESCRIPTION
**Pull request by msv-moxis**
Definition Of Done
(Overlined = N/A)

### General
- [ ] Are all cases in the issue description fixed and fulfilled?
- [ ] Are all PRs linked on the issue (all Moodle versions)?

### Have you updated/created:
- [ ] Github actions and Codacy files in .github
- [ ] README.md file (Description / Global settings / Instance settings / Capabilities / Setup / Release notes)
- [ ] Version file (Version / Date / Dependencies)
- [ ] install.xml, install.php, upgrade.php, uninstall.php under db directory
- [ ] Language strings (en / da)
- [ ] [Privacy provider](https://moodledev.io/docs/apis/subsystems/privacy) (GDPR)
- [ ] Backup/Restore routine
- [ ] HTTP request files for PHPStorm HTTP Client (If needed)

### Have tests been added for the following:
- [ ] Business logic (Unit / Integration)
- [ ] Code coverage (Which covers Business logic)

### Code Standards & Considerations:
- [ ] All TODO / PTODO tags should be created as issues, if not fixed in the current issue
- [ ] Security - The use of sesskey, require login, (defined('MOODLE_INTERNAL') || die;)
- [ ] Adheres to PSR-12 and Codacy standards
- [ ] Is clear and easy to understand
- [ ] Minimizes performance impacts
- [ ] Utilizes capabilities where applicable
- [ ] Defines return values and property types
- [ ] Methods are named based on their functionality

### CODE REVIEW

File: tests/integration/app/backup/backup_settings_helper_test.php [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            $this->assertEquals('No modules to include in section.',$e->getMessage());
+            throw new \Exception('');
+        }
+    }
+
+    public function test_get_course_settings_by_item_course_with_no_modules(): void
+    {
+        // Edge case
+        // Adhoc task runs after all modules in course have been deleted
+        $item = new entity();
```
Comment (frederikmillingpytlick) [inline tests/integration/app/backup/backup_settings_helper_test.php:273]: use factory

File: tests/integration/app/backup/backup_settings_helper_test.php [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        $item->set_type('section');
+        $item->set_old_instance_id($this->module1->cmid);
+
+        $this->expectException(\Exception::class);
+        try {
+            $this->helper->get_course_settings_by_item($item, true);
+        }catch (\Exception $e){
+            $this->assertEquals('No sections found by section id: ' . $this->module1->cmid ,$e->getMessage());
+            throw new \Exception('');
+        }
```
Comment (frederikmillingpytlick) [inline tests/integration/app/backup/backup_settings_helper_test.php:231]: Perhaps use `$this->expectException(...);`

**Review by frederikmillingpytlick — CHANGES_REQUESTED** [review]
After reviewing i found that the base factory doesn't have any replacements for the globals. - You can just add those to the base factory

File: classes/app/backup/backup_settings_helper.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        $settings += $this->get_module_settings($modules, $section_id, $module_id, $users);
+
+        $settings += $this->get_section_settings($sections, $section_id, $users);
+
+        return $settings;
+    }
+
+    private function get_ids_by_item(entity $item): array
+    {
+        $module_id = null;
```
[the remark this answers is not in the document]
Comment (frederikmillingpytlick) [reply classes/app/backup/backup_settings_helper.php:40]: Not resolved

File: classes/app/backup/backup_settings_helper.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    private function get_course_module_settings(
+        array $course_modules,
+        int $section_id,
+        ?int $course_module_id,
+        bool $include_users
+    ): array
+    {
+        $settings = [];
+        foreach ($course_modules as $module) {
```
Comment (frederikmillingpytlick) [inline classes/app/backup/backup_settings_helper.php:111]:
```suggestion
        foreach ($course_modules as $course_module) {
```

**Review by frederikmillingpytlick — CHANGES_REQUESTED** [review]
No need to increase version in last commit. Either update the version in `db/upgrade.php` or revert it in `version.php`

**Review by frederikmillingpytlick — APPROVED** [review]

File: classes/task/asynchronous_backup_task.php [hunk 5]
```diff
@@ -130,7 +132,8 @@ private function before_backup_started_hook(\backup_controller $backup_controlle
                 'grade_histories' => false,
                 'users' => false,
                 'anonymize' => false,
-                'include_badges' => false
+                'include_badges' => false,
+                'filename' => 'backup-'. $item_entity->get_type() . '_' . $item_entity->get_timecreated() . '.mbz'
```
Comment (frederikmillingpytlick) [inline classes/task/asynchronous_backup_task.php:136]: I just realized we could also just use the sharing cart item id. It's completely unique and a simple method of ensuring files don't collide
```suggestion
                'filename' => 'sharing_cart_backup-'. $item_entity->get_id() . '.mbz'
```

**Review by frederikmillingpytlick — CHANGES_REQUESTED** [review]

**Review by frederikmillingpytlick — APPROVED** [review]
Looks good :)

**Review by frederikmillingpytlick — APPROVED** [review]
